temporalio/bridge/worker.py

```diff
@@ -86,7 +91,137 @@ class FixedSizeSlotSupplier:
     num_slots: int
 
 
-SlotSupplier: TypeAlias = Union[FixedSizeSlotSupplier, ResourceBasedSlotSupplier]
+class SlotPermit:
+    """A permit to use a slot for a workflow/activity/local activity task.
+
+    You can inherit from this class to add your own data to the permit.
+    """
+
+    pass
+
+
+# WARNING: This must match Rust worker::SlotReserveCtx
+class SlotReserveContext(Protocol):
+    """Context for reserving a slot from a :py:class:`CustomSlotSupplier`."""
+
+    slot_type: Literal["workflow", "activity", "local-activity"]
+    """The type of slot trying to be reserved. Always one of "workflow", "activity", or "local-activity"."""
+    task_queue: str
+    """The name of the task queue for which this reservation request is associated."""
+    worker_identity: str
+    """The identity of the worker that is requesting the reservation."""
+    worker_build_id: str
+    """The build id of the worker that is requesting the reservation."""
+    is_sticky: bool
+    """True iff this is a reservation for a sticky poll for a workflow task."""
+
+
+@runtime_checkable
+class WorkflowSlotInfo(Protocol):
+    """Info about a workflow task slot usage."""
+
+    workflow_type: str
+    is_sticky: bool
+
+
+@runtime_checkable
+class ActivitySlotInfo(Protocol):
+    """Info about an activity task slot usage."""
+
+    activity_type: str
+
+
+@runtime_checkable
+class LocalActivitySlotInfo(Protocol):
+    """Info about a local activity task slot usage."""
+
+    activity_type: str
+
+
+SlotInfo: TypeAlias = Union[WorkflowSlotInfo, ActivitySlotInfo, LocalActivitySlotInfo]
+
+
+# WARNING: This must match Rust worker::SlotMarkUsedCtx
+class SlotMarkUsedContext(Protocol):
+    """Context for marking a slot used from a :py:class:`CustomSlotSupplier`."""
+
+    slot_info: SlotInfo
+    """Info about the task that will be using the slot."""
+    permit: SlotPermit
+    """The permit that was issued when the slot was reserved."""
+
+
+@dataclass(frozen=True)
+class SlotReleaseContext:
+    """Context for releasing a slot from a :py:class:`CustomSlotSupplier`."""
+
+    slot_info: SlotInfo
+    """Info about the task that will be using the slot."""
+    permit: SlotPermit
+    """The permit that was issued when the slot was reserved."""
+
+
+class CustomSlotSupplier(ABC):
```

Comment:
Are we still marking these as experimental?

Reply:
Will do